### PR TITLE
Change run_jobflow in EMR connection to use regional debugging jar wh…

### PR DESCRIPTION
**Pull request by louisdang**
…en enable_debugging is on. Fixes bug where jobs fail in ap-northeast-2 and eu-central-1.

### CODE REVIEW

**Comment by louisdang** [comment]
tox output:

louidang@a82066444bb3 ~/w/boto> tox
py26 installed: DEPRECATION: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of pip will drop support for Python 2.6,argparse==1.2.1,coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,linecache2==1.0.0,mock==1.0.1,nose==1.3.3,ordereddict==1.1,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5,six==1.10.0,traceback2==1.4.0,unittest2==1.1.0
py26 runtests: PYTHONHASHSEED='0'
py26 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.104s

OK (SKIP=4)
py27 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py27 runtests: PYTHONHASHSEED='0'
py27 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.752s

OK (SKIP=4)
py33 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py33 runtests: PYTHONHASHSEED='3299149326'
py33 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.764s

OK (SKIP=4)
py34 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py34 runtests: PYTHONHASHSEED='3299149326'
py34 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.216s

OK (SKIP=4)
pypy installed: /Users/louidang/workspaces/boto/.tox/pypy/site-packages/pip/_vendor/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.,  warnings.warn(msg),cffi==0.8.6,coverage==3.7.1,ecdsa==0.13,greenlet==0.4.0,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,readline==6.2.4.1,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
pypy runtests: PYTHONHASHSEED='0'
pypy runtests: commands[0] | python tests/test.py default
/Users/louidang/workspaces/boto/.tox/pypy/site-packages/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.
  warnings.warn(msg)
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..E.....................................................................................................................................................................................................................................................................................................E.....S..........................................E................................................................................................................................................................................................................................................................................................S................................................................................................................
======================================================================
ERROR: test_decoding_to_dynamodb (tests.unit.dynamodb.test_types.TestDynamizer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/dynamodb/test_types.py", line 64, in test_decoding_to_dynamodb
    self.assertEqual(dynamizer.decode({'B': 'AQ=='}), types.Binary(b'\x01'))
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 347, in decode
    return decoder(attr[dynamodb_type])
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 362, in _decode_b
    return convert_binary(attr)
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 99, in convert_binary
    return Binary(base64.b64decode(n))
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
TypeError: expected bytes, not str

======================================================================
ERROR: test_resume_archive_from_file (tests.unit.glacier.test_layer2.TestVault)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/.tox/pypy/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/Users/louidang/workspaces/boto/tests/unit/glacier/test_layer2.py", line 293, in test_resume_archive_from_file
    sentinel.upload_id, file_obj=sentinel.file_obj)
  File "/Users/louidang/workspaces/boto/boto/glacier/vault.py", line 229, in resume_archive_from_file
    part_tree_hash = codecs.decode(part_desc['SHA256TreeHash'], 'hex_codec')
  File "/Users/louidang/workspaces/boto/.tox/pypy/lib-python/3/encodings/hex_codec.py", line 20, in hex_decode
    return (binascii.a2b_hex(input), len(input))
TypeError: 'str' does not support the buffer interface

======================================================================
ERROR: test_get_credential_report (tests.unit.iam.test_connection.TestGetCredentialReport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/iam/test_connection.py", line 362, in test_get_credential_report
    ['content'])
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
nose.proxy.TypeError: expected bytes, not str
-------------------- >> begin captured logging << --------------------
boto: DEBUG: Using access key provided by client.
boto: DEBUG: Using secret key provided by client.
boto: DEBUG: Method: POST
boto: DEBUG: Path: /
boto: DEBUG: Data:
boto: DEBUG: Headers: {}
boto: DEBUG: Host: iam.amazonaws.com
boto: DEBUG: Port: 443
boto: DEBUG: Params: {'Action': 'GetCredentialReport', 'Version': '2010-05-08'}
boto: DEBUG: establishing HTTPS connection: host=iam.amazonaws.com, kwargs={'port': 443, 'timeout': 70}
boto: DEBUG: Token: None
boto: DEBUG: CanonicalRequest:
POST
/

host:iam.amazonaws.com
x-amz-date:20160310T001925Z

host;x-amz-date
fd3045db4acd629ed8a5fe7042ae7d75e3fc53c19684230fd6431337134051cd
boto: DEBUG: StringToSign:
AWS4-HMAC-SHA256
20160310T001925Z
20160310/us-east-1/iam/aws4_request
3eecae8e5d810a15f8965d7508b78fe61771f4a29d044621c8bec08255a7dc8a
boto: DEBUG: Signature:
091c532c298a6932511e54b3f7a3e6931f84b09ad6f7502d531a20a4943839a4
boto: DEBUG: Final headers: {'Content-Length': '45', 'User-Agent': 'Boto/2.39.0 Python/3.2.5 Darwin/14.5.0', 'Host': 'iam.amazonaws.com', 'X-Amz-Date': '20160310T001925Z', 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8', 'Authorization': 'AWS4-HMAC-SHA256 Credential=aws_access_key_id/20160310/us-east-1/iam/aws4_request,SignedHeaders=host;x-amz-date,Signature=091c532c298a6932511e54b3f7a3e6931f84b09ad6f7502d531a20a4943839a4'}
boto: DEBUG: Response headers: []
boto: DEBUG: b'\n          <GetCredentialReportResponse>\n            <ResponseMetadata>\n              <RequestId>99e60e9a-0db5-11e4-94d4-b764EXAMPLE</RequestId>\n            </ResponseMetadata>\n            <GetCredentialReportResult>\n              <Content>RXhhbXBsZQ==</Content>\n              <ReportFormat>text/csv</ReportFormat>\n              <GeneratedTime>2014-07-17T11:09:11Z</GeneratedTime>\n            </GetCredentialReportResult>\n          </GetCredentialReportResponse>\n        '
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 964 tests in 10.251s

FAILED (SKIP=4, errors=3)
ERROR: InvocationError: '/Users/louidang/workspaces/boto/.tox/pypy/bin/python tests/test.py default'
__________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________
  py26: commands succeeded
  py27: commands succeeded
  py33: commands succeeded
  py34: commands succeeded
ERROR:   pypy: commands failed

**Comment by louisdang** [comment]
```
 louidang@a82066444bb3 ~/w/boto> tox
py26 installed: DEPRECATION: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of pip will drop support for Python 2.6,argparse==1.2.1,coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,linecache2==1.0.0,mock==1.0.1,nose==1.3.3,ordereddict==1.1,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5,six==1.10.0,traceback2==1.4.0,unittest2==1.1.0
py26 runtests: PYTHONHASHSEED='0'
py26 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.104s

OK (SKIP=4)
py27 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py27 runtests: PYTHONHASHSEED='0'
py27 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.752s

OK (SKIP=4)
py33 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py33 runtests: PYTHONHASHSEED='3299149326'
py33 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.764s

OK (SKIP=4)
py34 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py34 runtests: PYTHONHASHSEED='3299149326'
py34 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.216s

OK (SKIP=4)
pypy installed: /Users/louidang/workspaces/boto/.tox/pypy/site-packages/pip/_vendor/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.,  warnings.warn(msg),cffi==0.8.6,coverage==3.7.1,ecdsa==0.13,greenlet==0.4.0,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,readline==6.2.4.1,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
pypy runtests: PYTHONHASHSEED='0'
pypy runtests: commands[0] | python tests/test.py default
/Users/louidang/workspaces/boto/.tox/pypy/site-packages/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.
  warnings.warn(msg)
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..E.....................................................................................................................................................................................................................................................................................................E.....S..........................................E................................................................................................................................................................................................................................................................................................S................................................................................................................
======================================================================
ERROR: test_decoding_to_dynamodb (tests.unit.dynamodb.test_types.TestDynamizer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/dynamodb/test_types.py", line 64, in test_decoding_to_dynamodb
    self.assertEqual(dynamizer.decode({'B': 'AQ=='}), types.Binary(b'\x01'))
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 347, in decode
    return decoder(attr[dynamodb_type])
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 362, in _decode_b
    return convert_binary(attr)
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 99, in convert_binary
    return Binary(base64.b64decode(n))
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
TypeError: expected bytes, not str

======================================================================
ERROR: test_resume_archive_from_file (tests.unit.glacier.test_layer2.TestVault)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/.tox/pypy/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/Users/louidang/workspaces/boto/tests/unit/glacier/test_layer2.py", line 293, in test_resume_archive_from_file
    sentinel.upload_id, file_obj=sentinel.file_obj)
  File "/Users/louidang/workspaces/boto/boto/glacier/vault.py", line 229, in resume_archive_from_file
    part_tree_hash = codecs.decode(part_desc['SHA256TreeHash'], 'hex_codec')
  File "/Users/louidang/workspaces/boto/.tox/pypy/lib-python/3/encodings/hex_codec.py", line 20, in hex_decode
    return (binascii.a2b_hex(input), len(input))
TypeError: 'str' does not support the buffer interface

======================================================================
ERROR: test_get_credential_report (tests.unit.iam.test_connection.TestGetCredentialReport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/iam/test_connection.py", line 362, in test_get_credential_report
    ['content'])
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
nose.proxy.TypeError: expected bytes, not str
-------------------- >> begin captured logging << --------------------
boto: DEBUG: Using access key provided by client.
boto: DEBUG: Using secret key provided by client.
boto: DEBUG: Method: POST
boto: DEBUG: Path: /
boto: DEBUG: Data:
boto: DEBUG: Headers: {}
boto: DEBUG: Host: iam.amazonaws.com
boto: DEBUG: Port: 443
boto: DEBUG: Params: {'Action': 'GetCredentialReport', 'Version': '2010-05-08'}
boto: DEBUG: establishing HTTPS connection: host=iam.amazonaws.com, kwargs={'port': 443, 'timeout': 70}
boto: DEBUG: Token: None
boto: DEBUG: CanonicalRequest:
POST
/

host:iam.amazonaws.com
x-amz-date:20160310T001925Z

host;x-amz-date
fd3045db4acd629ed8a5fe7042ae7d75e3fc53c19684230fd6431337134051cd
boto: DEBUG: StringToSign:
AWS4-HMAC-SHA256
20160310T001925Z
20160310/us-east-1/iam/aws4_request
3eecae8e5d810a15f8965d7508b78fe61771f4a29d044621c8bec08255a7dc8a
boto: DEBUG: Signature:
091c532c298a6932511e54b3f7a3e6931f84b09ad6f7502d531a20a4943839a4
boto: DEBUG: Final headers: {'Content-Length': '45', 'User-Agent': 'Boto/2.39.0 Python/3.2.5 Darwin/14.5.0', 'Host': 'iam.amazonaws.com', 'X-Amz-Date': '20160310T001925Z', 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8', 'Authorization': 'AWS4-HMAC-SHA256 Credential=aws_access_key_id/20160310/us-east-1/iam/aws4_request,SignedHeaders=host;x-amz-date,Signature=091c532c298a6932511e54b3f7a3e6931f84b09ad6f7502d531a20a4943839a4'}
boto: DEBUG: Response headers: []
boto: DEBUG: b'\n          <GetCredentialReportResponse>\n            <ResponseMetadata>\n              <RequestId>99e60e9a-0db5-11e4-94d4-b764EXAMPLE</RequestId>\n            </ResponseMetadata>\n            <GetCredentialReportResult>\n              <Content>RXhhbXBsZQ==</Content>\n              <ReportFormat>text/csv</ReportFormat>\n              <GeneratedTime>2014-07-17T11:09:11Z</GeneratedTime>\n            </GetCredentialReportResult>\n          </GetCredentialReportResponse>\n        '
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 964 tests in 10.251s

FAILED (SKIP=4, errors=3)
ERROR: InvocationError: '/Users/louidang/workspaces/boto/.tox/pypy/bin/python tests/test.py default'
__________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________
  py26: commands succeeded
  py27: commands succeeded
  py33: commands succeeded
  py34: commands succeeded
ERROR:   pypy: commands failed 
```

File: tests/unit/emr/test_connection.py [hunk 1]
```diff
@@ -997,8 +1030,25 @@ def test_run_jobflow_service_role(self):
...
+        actual_params = self.actual_request.params.copy().items()
+
+        expected_params = set([
+            ('Steps.member.1.HadoopJarStep.Jar',
+             's3://ap-northeast-2.elasticmapreduce/libs/script-runner/script-runner.jar'),
+            ('Steps.member.1.HadoopJarStep.Args.member.1',
+                's3://ap-northeast-2.elasticmapreduce/libs/state-pusher/0.1/fetch'),
+        ])
+
+        self.assertTrue(expected_params <= actual_params)
```
Comment (kyleknap) [inline tests/unit/emr/test_connection.py]: Is there a reason that you did not use ``assert_request_parameters`` or ``assertEqual``? I would prefer that check as opposed to a length check.

**Comment by kyleknap** [comment]
Changes look fine to me. I am not sure why your pypy tox environment is not working, but Jenkins is passing which is good. Just had one question about the test you added.

File: tests/unit/emr/test_connection.py (hunk 1 above)
Comment (louisdang) [inline tests/unit/emr/test_connection.py]: I have to double check, but my reasoning was that assert_request_parameters test for list equality but I don't think this test should fail if the ordering of the params changes or if someone adds some new params to run_jobflow. 

Also, this isn't intended as length check, but a subset check (expected_params should be a subset of actual_params). I think I forgot to convert actual_params to a set, so I will fix this.

Comment (kyleknap) [inline tests/unit/emr/test_connection.py]: Ahh makes sense if it is a subset comparison. That would be cool if you can fix that.

**Comment by louisdang** [comment]
Fixed ^_^. I will paste the new tox output in a separate message.

**Comment by louisdang** [comment]
tox output:

```
py26 installed: DEPRECATION: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of pip will drop support for Python 2.6,argparse==1.2.1,coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,linecache2==1.0.0,mock==1.0.1,nose==1.3.3,ordereddict==1.1,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5,six==1.10.0,traceback2==1.4.0,unittest2==1.1.0
py26 runtests: PYTHONHASHSEED='0'
py26 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.166s

OK (SKIP=4)
py27 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py27 runtests: PYTHONHASHSEED='0'
py27 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
.....................................................................................................................................................................................................................S...S...............................................................................................................................................................................................................................................................................................................S.........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 4.917s

OK (SKIP=4)
py33 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py33 runtests: PYTHONHASHSEED='1291498838'
py33 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 4.719s

OK (SKIP=4)
py34 installed: coverage==3.7.1,ecdsa==0.13,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
py34 runtests: PYTHONHASHSEED='1291498838'
py34 runtests: commands[0] | python tests/test.py default
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..............................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................................................................S................................................................................................................
----------------------------------------------------------------------
Ran 964 tests in 5.437s

OK (SKIP=4)
pypy installed: /Users/louidang/workspaces/boto/.tox/pypy/site-packages/pip/_vendor/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.,  warnings.warn(msg),cffi==0.8.6,coverage==3.7.1,ecdsa==0.13,greenlet==0.4.0,httpretty==0.8.6,mock==1.0.1,nose==1.3.3,paramiko==1.16.0,pyasn1==0.1.9,pycrypto==2.6.1,PyYAML==3.11,readline==6.2.4.1,requests==2.0.1,rsa==3.1.4,simplejson==3.6.5
pypy runtests: PYTHONHASHSEED='0'
pypy runtests: commands[0] | python tests/test.py default
/Users/louidang/workspaces/boto/.tox/pypy/site-packages/pkg_resources/__init__.py:87: UserWarning: Support for Python 3.0-3.2 has been dropped. Future versions will fail here.
  warnings.warn(msg)
nose command: tests/test.py -a !notdefault tests/unit
....................................................................................................................................................................................................................S...S..E.....................................................................................................................................................................................................................................................................................................E.....S..........................................E................................................................................................................................................................................................................................................................................................S................................................................................................................
======================================================================
ERROR: test_decoding_to_dynamodb (tests.unit.dynamodb.test_types.TestDynamizer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/dynamodb/test_types.py", line 64, in test_decoding_to_dynamodb
    self.assertEqual(dynamizer.decode({'B': 'AQ=='}), types.Binary(b'\x01'))
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 347, in decode
    return decoder(attr[dynamodb_type])
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 362, in _decode_b
    return convert_binary(attr)
  File "/Users/louidang/workspaces/boto/boto/dynamodb/types.py", line 99, in convert_binary
    return Binary(base64.b64decode(n))
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
TypeError: expected bytes, not str

======================================================================
ERROR: test_resume_archive_from_file (tests.unit.glacier.test_layer2.TestVault)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/.tox/pypy/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/Users/louidang/workspaces/boto/tests/unit/glacier/test_layer2.py", line 293, in test_resume_archive_from_file
    sentinel.upload_id, file_obj=sentinel.file_obj)
  File "/Users/louidang/workspaces/boto/boto/glacier/vault.py", line 229, in resume_archive_from_file
    part_tree_hash = codecs.decode(part_desc['SHA256TreeHash'], 'hex_codec')
  File "/Users/louidang/workspaces/boto/.tox/pypy/lib-python/3/encodings/hex_codec.py", line 20, in hex_decode
    return (binascii.a2b_hex(input), len(input))
TypeError: 'str' does not support the buffer interface

======================================================================
ERROR: test_get_credential_report (tests.unit.iam.test_connection.TestGetCredentialReport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/louidang/workspaces/boto/tests/unit/iam/test_connection.py", line 362, in test_get_credential_report
    ['content'])
  File "/Users/louidang/.pyenv/versions/pypy3-2.4.0/lib-python/3/base64.py", line 83, in b64decode
    raise TypeError("expected bytes, not %s" % s.__class__.__name__)
nose.proxy.TypeError: expected bytes, not str
-------------------- >> begin captured logging << --------------------
boto: DEBUG: Using access key provided by client.
boto: DEBUG: Using secret key provided by client.
boto: DEBUG: Method: POST
boto: DEBUG: Path: /
boto: DEBUG: Data:
boto: DEBUG: Headers: {}
boto: DEBUG: Host: iam.amazonaws.com
boto: DEBUG: Port: 443
boto: DEBUG: Params: {'Action': 'GetCredentialReport', 'Version': '2010-05-08'}
boto: DEBUG: establishing HTTPS connection: host=iam.amazonaws.com, kwargs={'port': 443, 'timeout': 70}
boto: DEBUG: Token: None
boto: DEBUG: CanonicalRequest:
POST
/

host:iam.amazonaws.com
x-amz-date:20160314T220007Z

host;x-amz-date
fd3045db4acd629ed8a5fe7042ae7d75e3fc53c19684230fd6431337134051cd
boto: DEBUG: StringToSign:
AWS4-HMAC-SHA256
20160314T220007Z
20160314/us-east-1/iam/aws4_request
44a5fe22a197878ad064b6b1f507c89ce25dab9dbfc7852ecf93a2cdd951b6b9
boto: DEBUG: Signature:
8eb0dacbd2e52b800cb6fe706063da8594c137f1ddbafd1deb924dcceaba61be
boto: DEBUG: Final headers: {'Content-Length': '45', 'User-Agent': 'Boto/2.39.0 Python/3.2.5 Darwin/14.5.0', 'Host': 'iam.amazonaws.com', 'X-Amz-Date': '20160314T220007Z', 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8', 'Authorization': 'AWS4-HMAC-SHA256 Credential=aws_access_key_id/20160314/us-east-1/iam/aws4_request,SignedHeaders=host;x-amz-date,Signature=8eb0dacbd2e52b800cb6fe706063da8594c137f1ddbafd1deb924dcceaba61be'}
boto: DEBUG: Response headers: []
boto: DEBUG: b'\n          <GetCredentialReportResponse>\n            <ResponseMetadata>\n              <RequestId>99e60e9a-0db5-11e4-94d4-b764EXAMPLE</RequestId>\n            </ResponseMetadata>\n            <GetCredentialReportResult>\n              <Content>RXhhbXBsZQ==</Content>\n              <ReportFormat>text/csv</ReportFormat>\n              <GeneratedTime>2014-07-17T11:09:11Z</GeneratedTime>\n            </GetCredentialReportResult>\n          </GetCredentialReportResponse>\n        '
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 964 tests in 9.325s

FAILED (SKIP=4, errors=3)
ERROR: InvocationError: '/Users/louidang/workspaces/boto/.tox/pypy/bin/python tests/test.py default'
______________________________________________________________ summary ______________________________________________________________
  py26: commands succeeded
  py27: commands succeeded
  py33: commands succeeded
  py34: commands succeeded
ERROR:   pypy: commands failed
```

**Comment by kyleknap** [comment]
Thanks! Merging.

**Pull request merged**